### PR TITLE
Support building with Java 21

### PR DESCRIPTION
**Pull request by findepi**
This is currently the latest Java LTS version, so we should support
building with it.

### CODE REVIEW

File: gradle/wrapper/gradle-wrapper.properties
Comment (findepi) [inline gradle/wrapper/gradle-wrapper.properties:5]: submitted separate as https://github.com/apache/iceberg/pull/10476

File: .github/workflows/java-ci.yml
Comment (findepi) [inline .github/workflows/java-ci.yml]: submitted separately as https://github.com/apache/iceberg/pull/10473

File: .java-version
Comment (findepi) [inline .java-version]: submitted separately as https://github.com/apache/iceberg/pull/10472

**Comment by findepi** [comment]
The build is green, ready for review, but still blocked by https://github.com/apache/iceberg/issues/10368#issuecomment-2161039351. Therefore keeping as a draft.

**Comment by findepi** [comment]
cc @rdblue @szehon-ho @Fokko @jbonofre

**Comment by jbonofre** [comment]
@findepi thanks for the PR. As you know, I'm working on revapi and gradle update.

Supporting JDK21 is great, but we might need to keep previous Java version for some engine/plugins (I'm thinking about hive or pig modules for instance). We can do a follow up PR to remove or update dependencies to "align" Java version.

**Comment by szehon-ho** [comment]
others here are more familiar with the build than I am, will let them comment.  Is this removing some java version?

Also, why all the code formatting, can we remove those from the pr for ease of review?

**Comment by findepi** [comment]
> others here are more familiar with the build than I am, will let them comment. Is this removing some java version?

no, i don't think so

> Also, why all the code formatting, curious if we can remove those from the pr for ease of review?

I had to update the spotless plugin and google format library to support Java 21.
This resulted in dropped support for Java 8 and one-time code formatting changes -- clearly the new formatter is not backwards compatible. 
I can extract those changes to separate PR after we decide what to do with https://github.com/apache/iceberg/pull/10473.
Without https://github.com/apache/iceberg/pull/10473, the formatting checks wouldn't run on CI at all, AFAIU.

File: gradle/wrapper/gradle-wrapper.properties
Comment (nastra) [inline gradle/wrapper/gradle-wrapper.properties:5]: this seems unrelated to running this with JDK21. We currently can't upgrade Gradle due to an issue with RevAPI

File: baseline.gradle
Comment (nastra) [inline baseline.gradle:66]: I don't think it's a good idea to upgrade this to the latest version. The issue is that spotless will produce different formatting results depending on which JDK it's been executed with. So as long as the project builds & releases with JDK8 we need to stay at the GJF version 1.7

File: aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogCommitFailure.java
Comment (nastra) [inline aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogCommitFailure.java:535]: the formatting changes are most likely a result of updating the GJF version (which we can't update as long as we build/release with JDK8)

File: .github/workflows/api-binary-compatibility.yml
Comment (nastra) [inline .github/workflows/api-binary-compatibility.yml:30]: seems unrelated to the scope of the PR

**Review by nastra — CHANGES_REQUESTED** [review]

File: gradle/wrapper/gradle-wrapper.properties
Comment (findepi) [reply gradle/wrapper/gradle-wrapper.properties:5]: it is related. gradle 8.1.1 cannot be run on JDK 21

i am aware of the dependency, see https://github.com/apache/iceberg/pull/10474#issuecomment-2161508864

File: baseline.gradle
Comment (findepi) [reply baseline.gradle:66]: Sooner or later we will add support for Java 21 and we will drop support for Java 8. These two things do not have to depend on each other. Sadly, there is no google java formatter version which can run under Java 21 and under Java 8, so we need to make a choice which Java versions we can check source code formatting with. We can either choose 8-17 or 11-21, but not the union. The 11-21 is an inevitable choice, unless we want to live in the past for ever.
Now, code formatting is very important for project day to day, but feels not directly related to releases. The change here doesn't affect how we build releases.

File: .github/workflows/api-binary-compatibility.yml
Comment (findepi) [reply .github/workflows/api-binary-compatibility.yml:30]: it's also extracted as https://github.com/apache/iceberg/pull/10485, see rationale there.

this PR changes build system (gradle version) and it would not be smart to assume workflows continue to work without running them

Of course, this doesn't prevent all problems -- revapi silently not working is a type of problem this doesn't prevent -- but at least prevents some and in that sense it is related to the scope of this PR.

File: aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogCommitFailure.java
Comment (findepi) [reply aws/src/integration/java/org/apache/iceberg/aws/glue/TestGlueCatalogCommitFailure.java:535]: > the formatting changes are most likely a result of updating the GJF version

yes, and that's why they come in separate commit

```
Update spotless to run on Java 21 

This involves code formatting changes to comply with updated formatting
library.
```


> (which we can't update as long as we build/release with JDK8)


this is discussed at https://github.com/apache/iceberg/pull/10474#discussion_r1639514171

File: baseline.gradle
Comment (nastra) [reply baseline.gradle:66]: Once Iceberg drops support for Java8 is when we will be able to update the GJF version. As I mentioned earlier, we pinned the version so that `spotlessApply` doesn't produce different formatting results depending on the JDK version it is executed with. Since we release for JDK8, we need to use the GJF version that works with JDK8.

The community is planning to drop JDK8 support (which only Hive depends on) with Iceberg 2.0, which is also when we'll be upgrading most likely to 11-21. That being said, we can only upgrade the GJF version once we also move to JDK 11-21

Comment (findepi) [reply baseline.gradle:66]: Thanks for rephrasing your initial comment for me. This reinforces my understanding.

>  so that spotlessApply doesn't produce different formatting results depending on the JDK version it is executed with.

I totally agree this wouldn't be acceptable! Do you maybe remember under what circumstances this can be observed?
In this PR I modified the CI scripts so that `./gradlew -DallVersions build -x test -x javadoc -x integrationTest` runs with java 11-21 and all runs agree on the desired formatting.  While I can see that formatting depends on GJF version, I don't see it being dependent on the JDK version it is executed with.

> Since we release for JDK8, we need to use the GJF version that works with JDK8.

I hear you on this, but would appreciate helping me understanding the why. I still fail to see it.
Is it because we expect Iceberg contributors to locally run JDK 8? I am concerned that JDK 8 is unsupported, unmaintained publicly and potentially not safe to run, so would prefer not to force contributors to use it, even though the **releases** target Java 8 compatibility.

Comment (nastra) [reply baseline.gradle:66]: When I added spotless + configured GJF it would produce different results depending on the JDK version that was used when executing `spotlessApply`. I typically use `sdkman` to switch JDK version and from what I remember is that when you run `spotlessApply` using JDK8 it would produce one formatting and a different formatting when using JDK11+ (which you can also see in this PR, where the formatting changed). Back then I looked into why it behaves the way it does and it was due to  different JDK versions might pick up a different GJF version and so we had to pin the version to the latest one that was supported by JDK8.


Typically whoever does a release is expected to run the release script using JDK8 locally btw. I hope that helps in understanding the context as to why we had to pin the GJF version.

Comment (findepi) [reply baseline.gradle:66]: thank you for your comment, @nastra !

> using JDK8 it would produce one formatting and a different formatting when using JDK11+ (which you can also see in this PR, where the formatting changed).

in this PR formatting changed, because GJF version changed

>  Back then I looked into why it behaves the way it does and it was due to different JDK versions might pick up a different GJF version and so we had to pin the version to the latest one that was supported by JDK8.

that explains everything, thanks!

Indeed, formatting is apparently dependent on GJF version and GJF version should be pinned. 

on the code from this PR i run the following
```
JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home ./gradlew spotlessApply
JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-17.jdk/Contents/Home ./gradlew spotlessApply
JAVA_HOME=/Library/Java/JavaVirtualMachines/zulu-21.jdk/Contents/Home ./gradlew spotlessApply
```

and none of these commands resulted in code changes. The formatting rules seem consistent and remain independent from the JDK version the build runs with.


> Typically whoever does a release is expected to run the release script using JDK8 locally btw.

This is what i was assuming, thanks for confirming this!

Note that with changes here, the build continues to work with JDK 8.
It just doesn't offer code style checks under JDK 8, but the person running release doesn't need them. The release goes off `main` branch where all the checks were already done by the CI. 
If a person running a release wanted to do some code changes before creating the release build -- which sounds wrong to me -- they would still need to re-run all other CI checks, which involve JDK versions 8, 11 and 17 (currently; and 21 should be added to the mix).

> I hope that helps in understanding the context as to why we had to pin the GJF version.

Absolutely! So with this PR, the GJF version remains pinned. I hope this clarifies the concerns.

**Comment by findepi** [comment]
https://github.com/apache/iceberg/pull/10477, https://github.com/apache/iceberg/pull/10482, https://github.com/apache/iceberg/pull/10485 were extracted from here. Will rebase this PR once these are merged.

**Comment by jbonofre** [comment]
@findepi thanks. FYI, I'm fixing the `gradle-revapi` plugin and Ed and I will do the release, then, I will rebase the Gradle 8.8 PR and we are good 😄

File: build.gradle
Comment (nastra) [inline build.gradle:90]: extra JVM args seem to be the same as needed for JDK17, so maybe just add a `JavaVersion.current() >= JavaVersion.VERSION_17` and set `project.ext.jdkVersion` via `JavaVersion.current().getMajorVersion()`

File: spark/v3.3/build.gradle
Comment (nastra) [inline spark/v3.3/build.gradle:63]: does this also apply for Scala 2.13?

Comment (findepi) [reply spark/v3.3/build.gradle:63]: apparently not, i think the Ci would detect that

**Comment by findepi** [comment]
Finally not draft anymore!

File: baseline.gradle
Comment (nastra) [inline baseline.gradle:50]: I don't think we can merge the PR without having spotless enabled unfortunately

File: build.gradle
Comment (nastra) [reply build.gradle:90]: yeah that's why I mentioned to set `project.ext.jdkVersion` via `JavaVersion.current().getMajorVersion()` for JDK17+

File: baseline.gradle
Comment (findepi) [reply baseline.gradle:50]: I wish this was avoidable. I do not know of a way to run same formatting rules under Java 8 and Java 21 and that's we support today. 
I do believe that being able to test and actually testing with 21 is incremental improvement over current state. I see value in adding even partial support for 21. In particular, this PR identifies a problem with scala 2.12.

File: build.gradle
Comment (findepi) [reply build.gradle:67]: missed that, thanks! updated

File: baseline.gradle
Comment (nastra) [reply baseline.gradle:50]: Can you bring this topic up on the Dev mailing list to see what others think about adding JDK21 support without being able to run spotless?

Comment (nastra) [reply baseline.gradle:50]: @jbonofre we can't upgrade the GJF as long as we support JDK8 because upgrading the GJF version will produce different results depending on the underlying JDK it is executed with and so CI actions (and locally) will fail.

Comment (jbonofre) [reply baseline.gradle:50]: That's my point: I would postpone the JDK21 update right now, and jump directly to JDK21 without supporting JDK8 anymore.
I think it will be very difficult to support both JDK21 and JDK8 from spotless/GJF standpoint. So, I would directly jump to new version when possible.

Comment (nastra) [reply baseline.gradle:50]: Dropping JDK8 support means dropping Hive support (which I'm totally fine with) but we need to properly communicate which Iceberg version will drop Hive support. My understanding was that we were planning to drop Hive support with Iceberg 2.0

Comment (jbonofre) [reply baseline.gradle:50]: Yes exactly. It's what I proposed while ago in the Iceberg 2.0 discussion thread. So, I propose:
- on 1.x (including 1.6.0 release currently in preparation), we keep JDK8
- for 2.x, we merge this PR and jump to JDK21

I can reply on the Iceberg 2.0 thread if you want or Piotr can start a new thread about JDK21 on Iceberg 2.0.
I would love to have feedback on the Iceberg 2.0 thread 😄

Comment (BsoBird) [reply baseline.gradle:50]: > Dropping JDK8 support means dropping Hive support (which I'm totally fine with) but we need to properly communicate which Iceberg version will drop Hive support. My understanding was that we were planning to drop Hive support with Iceberg 2.0

Sir. Maybe the iceberg-hive module should be maintained by the hive community in the hive project. Because hive 3.X will elo sooner or later.

Comment (nastra) [reply baseline.gradle:50]: @BsoBird newer versions of Hive are maintained by the Hive community but the Iceberg community still maintains an older version of Hive

File: .github/workflows/spark-ci.yml
Comment (pan3793) [inline .github/workflows/spark-ci.yml:76]: Actually Spark 3.5 does not fully support Java 21, the right ticket is https://issues.apache.org/jira/browse/SPARK-43831

Comment (findepi) [reply .github/workflows/spark-ci.yml:76]: Thanks @pan3793 for your comment and the link. This is good reference.
If our tests with Spark 3.5 pass with Java 21, we don't need to disable them, but it's good to know we should be expecting challenges. I will rework the comment.

**Comment by findepi** [comment]
Since spotless plugin is not installed on JDK 21, ./gradlew spotlessApply would fail with

```
$ JAVA_HOME=~/.jenv/versions/21 ./gradlew spotlessApply

FAILURE: Build failed with an exception.

* What went wrong:
Task 'spotlessApply' not found in root project 'iceberg' and its subprojects.

* Try:
> Run gradlew tasks to get a list of available tasks.
> For more on name expansion, please refer to https://docs.gradle.org/8.8/userguide/command_line_interface.html#sec:name_abbreviation in the Gradle documentation.
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.8/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 14s
```

Added spotlessApply stub, for more actionable message:

```
$ JAVA_HOME=~/.jenv/versions/21 ./gradlew spotlessApply

FAILURE: Build failed with an exception.

* Where:
Script '/Users/findepi/repos/iceberg/baseline.gradle' line: 51

* What went wrong:
A problem occurred evaluating script.
> Spotless plugin is currently disabled when running on JDK 21 (until we drop JDK 8). To run spotlessApply please use a different JDK version.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.8/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 3s
```

File: baseline.gradle
Comment (nastra) [inline baseline.gradle:52]: thanks for adding that msg, that's going to be helpful

File: spark/v3.3/build.gradle
Comment (nastra) [reply spark/v3.3/build.gradle:63]: do you remember what the exact error msg was here?

**Review by nastra — APPROVED** [review]

**Review by Fokko — APPROVED** [review]

File: spark/v3.3/build.gradle
Comment (findepi) [reply spark/v3.3/build.gradle:63]: i don't, i should have kept a note

**Comment by nastra** [comment]
thanks for working on this @findepi

**Comment by findepi** [comment]
thanks for the merge!

**Pull request merged**
